projectq/cengines/_replacer/_decompositionhandling.py

```diff
@@ -11,67 +11,101 @@
 #   limitations under the License.
 
 from ._decomposition import Decomposition
+from projectq.ops import BasicGate
 
-decompositions = dict()
 
-
-class ThisIsNotAGateClassError(Exception):
+class ThisIsNotAGateClassError(TypeError):
     pass
 
 
-def register_decomposition(gate_class, gate_decomposer, gate_recognizer=None):
+class DecompositionRule:
```

Comment:
`DecompositionRule` seems very similar to `Decomposition`, so I'm wondering if we want to either merge the two or make `all_defined_decomposition_rules`, e.g., a dict of lists of Decompositions, so we can get the gate class when iterating through `all_defined_decomposition_rules`.